Bump cabal-bootstrap-gen to GHC 9.4.4 ecosystem

nmeum opened the pull request:
I updated the dependencies in the .cabal bootstrap file and sightly adjusted src/Main.hs for the `I.riCabal → I.riCabalHash` change. I successfully bootstrapped Cabal on Alpine Linux Edge with these changes. I am new to Cabal so let me know if this needs any adjustments.

Fixes #8613

---

Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

nmeum commented on bootstrap/src/Main.hs:
```diff
@@ -93,7 +93,7 @@ main2 meta plan = do
                             ( Hackage
                             , Just $ fromIntegral (I.riRevision relInfo)
-                            , P.sha256FromByteString $ I.getSHA256 $ I.riCabal relInfo
+                            , P.sha256FromByteString $ I.getSHA256 $ I.riCabalHash relInfo
```
This probably needs some sort of `#ifdef` in order to not break bootstrapping with GHC `<9.4.0`.

Mikolaj replied:
@nmeum: thank you! Yes, I guess it does. We'll need to test this with older GHC and tweak as needed.

nmeum replied:
It seems this change was introduced in https://github.com/haskell-CI/haskell-ci/commit/4c28d9961d5d3e69c84205693824cd62fb9a7672 with cabal-install-parsers-0.6. I added a corresponding `#if MIN_VERSION` check.

Mikolaj commented:
@nmeum: BTW, please feel free to compare with old PRs that update the generator and/or generate files for new GHCs.

nmeum commented:
> @nmeum: BTW, please feel free to compare with old PRs that update the generator and/or generate files for new GHCs.

I tried to follow #8403 but somehow this PR didn't include changes to the top-level `Makefile` and the `CI` I just updated the PR in the hope that I modified these files correctly.

robx reviewed (approved):
This looks good to me, though I haven't tested it locally. Bootstrap CI passes so the generated files should be fine at least. (CI doesn't cover `bootstrap-gen` itself.)

Mikolaj commented:
Let me rebase so that CI unbreaks (we worked around some infrastructure breakage).

@nmeum: did you perchance test the new scripts with GHCs other than 9.4.4? Ideally with all listed in the Makefile? I guess it's fine if they produce differenct bootstrap files than in the repo, but a smoke test would be comparing if the sizes are similar.

After your are satisfied about your due diligence, please kindly set the merge_me label.

Mikolaj commented:
@mergify rebase

Mikolaj commented:
@mergify rebase

nmeum commented:
Given that this passes CI and has been approved: Do you need me to do anything else to get this merged or is it just a question of time? :)

Kleidukos commented:
@nmeum All good on our side, you may set the label `merge-me` and remove the `attention: needs-review` label :)

Mikolaj commented:
@nmeum: as as @Kleidukos said. And feel free to read CONTRIBUTING.md if in doubt.

BTW, did you catch the below comment of mine?

> @nmeum: did you perchance test the new scripts with GHCs other than 9.4.4? Ideally with all listed in the Makefile? I guess it's fine if they produce differenct bootstrap files than in the repo, but a smoke test would be comparing if the sizes are similar.
> 
> After your are satisfied about your due diligence, please kindly set the merge_me label.

nmeum commented:
> BTW, did you catch the below comment of mine?
> 
> > @nmeum: did you perchance test the new scripts with GHCs other than 9.4.4? Ideally with all listed in the Makefile? I guess it's fine if they produce differenct bootstrap files than in the repo, but a smoke test would be comparing if the sizes are similar.

Sorry, I somehow missed your comment. I only tested this with GHC 9.4.4. I am using this changeset for bootstrapping Cabal for the [Alpine Linux](https://alpinelinux.org) [Cabal package](https://gitlab.alpinelinux.org/alpine/aports/-/tree/master/community/cabal-stage0) and hence have only tested this with the GHC version available in the Alpine repositories (9.4.4).

Mikolaj commented:
> I only tested this with GHC 9.4.4. I am using this changeset for bootstrapping Cabal for the [Alpine Linux](https://alpinelinux.org) [Cabal package](https://gitlab.alpinelinux.org/alpine/aports/-/tree/master/community/cabal-stage0) and hence have only tested this with the GHC version available in the Alpine repositories (9.4.4).

That's fine. If somebody manages to test during the 2 days delay period after setting the label, it'd be great. But if not, we still have a long time before cabal 3.10.2 or 3.12, so we may hear from other boostrappers and fix, as necessary.

Mikolaj commented:
@mergify backport 3.10
